fuel-block-producer/src/block_producer.rs

```diff
@@ -0,0 +1,194 @@
+use crate::{
+    block_producer::transaction_selector::select_transactions,
+    db::BlockProducerDatabase,
+    ports::{
+        Relayer,
+        TxPool,
+    },
+    Config,
+};
+use anyhow::{
+    Context,
+    Result,
+};
+use chrono::Utc;
+use fuel_core_interfaces::{
+    block_producer::{
+        BlockProducer as Trait,
+        Error::{
+            GenesisBlock,
+            InvalidDaFinalizationState,
+            MissingBlock,
+        },
+    },
+    common::{
+        crypto::ephemeral_merkle_root,
+        fuel_tx::CheckedTransaction,
+        fuel_types::Bytes32,
+    },
+    executor::{
+        Error,
+        ExecutionMode,
+        Executor,
+    },
+    model::{
+        BlockHeight,
+        DaBlockHeight,
+        FuelBlock,
+        FuelBlockHeader,
+    },
+};
+use tracing::{
+    debug,
+    error,
+};
+
+#[cfg(test)]
+mod tests;
+mod transaction_selector;
+
+pub struct Producer<'a> {
+    pub config: Config,
+    pub db: &'a dyn BlockProducerDatabase,
+    pub txpool: &'a dyn TxPool,
+    pub executor: &'a dyn Executor,
+    pub relayer: &'a dyn Relayer,
+}
+
+#[async_trait::async_trait]
+impl<'a> Trait for Producer<'a> {
+    /// Produces a block for the specified height
+    async fn produce_block(&self, height: BlockHeight) -> Result<FuelBlock> {
+        //  - get previous block info (hash, root, etc)
+        //  - select best da_height from relayer
+        //  - get available txs from txpool
+        //  - select best txs based on factors like:
+        //      1. fees
+        //      2. parallel throughput
+        //  - Execute block with production mode to correctly malleate txs outputs and block headers
+
+        let previous_block_info = self.previous_block_info(height).await?;
```

Comment:
Seems `previous_block_info` is sync method

Reply:
fixed